tests/ws.c

```diff
@@ -29,15 +30,58 @@
 
 #include "testutil.h"
 
+#define SOCKET_ADDRESS  "ws://127.0.0.1:5555"
+
 /*  Basic tests for WebSocket transport. */
 
+/*  test_text() verifies that we drop messages properly when sending invalid
+    UTF-8, but not when we send valid data. */
+void test_text() {
+
+    int sb;
+    int sc;
+    int opt;
+    uint8_t bad[20];
+
+    /*  Negative testing... bad UTF-8 data for text. */
+    sb = test_socket (AF_SP, NN_PAIR);
+    sc = test_socket (AF_SP, NN_PAIR);
+
+    /*  Wait for connects to establish. */
+    nn_sleep (200);
+
+    opt = NN_WS_MSG_TYPE_TEXT;
+    test_setsockopt(sb, NN_WS, NN_WS_MSG_TYPE, &opt, sizeof (opt));
+    opt = NN_WS_MSG_TYPE_TEXT;
+    test_setsockopt(sc, NN_WS, NN_WS_MSG_TYPE, &opt, sizeof (opt));
+    opt = 500;
+    test_setsockopt(sb, NN_SOL_SOCKET, NN_RCVTIMEO, &opt, sizeof (opt));
+
+    test_bind (sb, SOCKET_ADDRESS);
+    test_connect (sc, SOCKET_ADDRESS);
+
+    test_send (sc, "GOOD");
+    test_recv (sb, "GOOD");
+
+    /*  and the bad ... */
+    strcpy((char *)bad, "BAD.");
+    bad[2] = (char)0xDD;
+    test_send (sc, (char *)bad);
+
+    /*  Make sure we dropped the frame.  Apparently it comes back as EAGAIN
+        rather than ETIMEDOUT.  I think this possibly a doc bug. */
```

Comment:
If its not evident, that’s a failure in my ability to express what is
happening.

The recv is timing out due to no input.  Because we drop the frame (and the
connection) when the invalid data is first received at the bottom transport
layer, without ever sending anything up to the application.   (No zero
length frame, nor any error.  The application is entirely shielded from
transport errors — this is part of the design principles of libnanomsg.)

Without a RCVTIMEO timer, the recv doesn’t timeout at all, but just blocks
forever.  I put the timeout in so that I could detect that the frame was
dropped without delivery.

Now it turns out that the code is buggy here, and returns EAGAIN instead of
ETIMEDOUT.  That’s a bug in the core socket handling code, and I plan to
fix it (its a trivial change) to make it conform to the documented
behavior.  EAGAIN is only for reporting an empty recv queue when using
non-blocking I/O.

Does that clear it up at all?

On Sat, Oct 24, 2015 at 7:57 PM, Wirebird Labs LLC <notifications@github.com
> wrote:

> In tests/ws.c
> <https://github.com/nanomsg/nanomsg/pull/492#discussion_r42940349>:
>
> > +    opt = 500;
> > +    test_setsockopt(sb, NN_SOL_SOCKET, NN_RCVTIMEO, &opt, sizeof (opt));
> > +
> > +    test_bind (sb, SOCKET_ADDRESS);
> > +    test_connect (sc, SOCKET_ADDRESS);
> > +
> > +    test_send (sc, "GOOD");
> > +    test_recv (sb, "GOOD");
> > +
> > +    /*  and the bad ... */
> > +    strcpy((char *)bad, "BAD.");
> > +    bad[2] = (char)0xDD;
> > +    test_send (sc, (char *)bad);
> > +
> > +    /*  Make sure we dropped the frame.  Apparently it comes back as EAGAIN
> > +        rather than ETIMEDOUT.  I think this possibly a doc bug. */
>
> would be at the sender side
>
> This is sender-side; line 69 above.
>
> That said ... I now see that sws.c does not validate that an outbound
> message "claimed" to be UTF-8 is indeed UTF-8 -- it only validates incoming
> (untrusted) messages. This seems reasonable.
>
> Meaning, of course, it could not return EINVAL or EPROTO without
> validating the claim of the application. That said ... it's still not
> evident why EAGAIN or ETIMEDOUT would be the expected error code ...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nanomsg/nanomsg/pull/492/files#r42940349>.
>